# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [6]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [7]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [8]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [9]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [10]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [11]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [12]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [13]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [14]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [15]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [16]:
pisa_clean.sample(10)

STIDSTD ST04Q01                                ST13Q01 ST14Q01  \
116665     2567  Female  She did not complete <ISCED level 1>       No   
175128     4233  Female                      <ISCED level 3A>      NaN   
115776     1678  Female                      <ISCED level 3A>      NaN   
255568    14979  Female                       <ISCED level 2>      NaN   
459319     4943  Female                       <ISCED level 1>       No   
150886     5304    Male                       <ISCED level 2>       No   
483973     3443  Female                       <ISCED level 1>       No   
182740     7066  Female                  <ISCED level 3B, 3C>       No   
182154     6480  Female                      <ISCED level 3A>       No   
182921     7247    Male                      <ISCED level 3A>       No   

       ST14Q02 ST14Q03 ST14Q04                ST17Q01 ST18Q01 ST18Q02  ...  \
116665     NaN     NaN     NaN       <ISCED level 1>       No     NaN  ...   
175128     Yes     NaN     NaN      <ISCED level 3A>      NaN     Yes  ...   
115776     NaN     Yes     NaN       <ISCED level 2>      NaN     NaN  ...   
255568     NaN     NaN     NaN       <ISCED level 2>      NaN     NaN  ...   
459319      No     NaN      No       <ISCED level 1>       No      No  ...   
150886      No      No      No      <ISCED level 3A>       No      No  ...   
483973      No     NaN      No       <ISCED level 1>       No      No  ...   
182740     Yes      No      No  <ISCED level 3B, 3C>       No     Yes  ...   
182154     Yes     Yes      No      <ISCED level 3A>       No     Yes  ...   
182921      No      No      No  <ISCED level 3B, 3C>       No      No  ...   

       ST57Q03 ST57Q04  ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
116665     NaN     NaN      NaN    NaN    NaN    NaN           244.98366   
175128     2.0     0.0      0.0  225.0  180.0  270.0           529.99686   
115776     NaN     NaN      NaN  120.0  100.0  120.0           356.76118   
255568     NaN     NaN      NaN    NaN    NaN    NaN           392.82596   
459319     NaN     NaN      NaN    NaN    NaN    NaN           398.04486   
150886     4.0     0.0      0.0  200.0  150.0    NaN           349.36128   
483973     0.0    10.0      0.0    NaN    NaN    NaN           573.46160   
182740     0.0     0.0      1.0  135.0  180.0  225.0           396.64280   
182154     NaN     NaN      NaN    NaN    NaN    NaN           568.00902   
182921     0.0     0.0      0.0  180.0  135.0  135.0           434.49916   

        Average Reading Score  Average Science Score  Average Total Score  
116665              332.58162              311.39500           377.981313  
175128              603.36190              592.44682           751.934147  
115776              398.42988              377.78816           496.580133  
255568              421.70320              410.61168           539.322267  
459319              454.42878              395.03914           548.519213  
150886              347.38438              366.13204           470.746327  
483973              543.15318              595.24430           761.773560  
182740              450.53664              458.91458           567.578940  
182154              539.02280              565.96418           747.001673  
182921              368.87646              446.88552           561.586767  

[10 rows x 24 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [17]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [18]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

# Null

In [19]:
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [20]:
father_edu = pisa_clean.copy()

In [21]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [22]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [23]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [24]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [25]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
159824       14242    Male           <ISCED level 1>    
140654        2553  Female      <ISCED level 3B, 3C>    
147325        1743    Male          <ISCED level 3A>    
264344       23755    Male           <ISCED level 2>    
203250        1475  Female          <ISCED level 3A>    
168617       23035    Male          <ISCED level 3A>    
83466         8997    Male          <ISCED level 3A>    
12297         7555  Female          <ISCED level 3A>    
463076        2094    Male          <ISCED level 3A>    
357590        3338    Male      <ISCED level 3B, 3C>    

       Mother Qualifications - ISCED level 6  \
159824                                    No   
140654                                    No   
147325                                    No   
264344                                    No   
203250                                    No   
168617                                    No   
83466                                     No   
12297                                     No   
463076                                    No   
357590                                    No   

       Mother Qualifications - ISCED level 5A  \
159824                                     No   
140654                                     No   
147325                                     No   
264344                                     No   
203250                                     No   
168617                                    Yes   
83466                                      No   
12297                                      No   
463076                                     No   
357590                                     No   

       Mother Qualifications - ISCED level 5B  \
159824                                     No   
140654                                     No   
147325                                     No   
264344                                     No   
203250                                     No   
168617                                     No   
83466                                      No   
12297                                      No   
463076                                    Yes   
357590                                     No   

       Mother Qualifications - ISCED level 4  \
159824                                    No   
140654                                    No   
147325                                    No   
264344                                    No   
203250                                    No   
168617                                    No   
83466                                     No   
12297                                     No   
463076                                    No   
357590                                    No   

                  Father - Highest Schooling  <ISCED level 6>  \
159824  He did not complete <ISCED level 1>                No   
140654                 <ISCED level 3B, 3C>                No   
147325                      <ISCED level 2>                No   
264344                      <ISCED level 2>                No   
203250                     <ISCED level 3A>                No   
168617                      <ISCED level 2>                No   
83466                      <ISCED level 3A>   <ISCED level 6>   
12297                      <ISCED level 3A>                No   
463076                     <ISCED level 3A>                No   
357590                 <ISCED level 3B, 3C>                No   

        <ISCED level 5>  ... Learning Time - Science Average Math Score  \
159824               No  ...                   200.0          471.49868   
140654               No  ...                   225.0          497.74888   
147325               No  ...                   330.0          669.58252   
264344               No  ...                   120.0          469.62926   
203250               No  ...                   180.0          353.17804   
168617               No  ...                   150.0          649.64170   
8

In [26]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [27]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
293942        2388  Female          <ISCED level 3A>    
302093       16910  Female           <ISCED level 2>    
195795        3392    Male      <ISCED level 3B, 3C>    
975147        4102  Female           <ISCED level 2>    
479944       13609  Female          <ISCED level 3A>    
290007        1394    Male          <ISCED level 3A>    
21021        13134    Male          <ISCED level 3A>    
948592         472  Female          <ISCED level 3A>    
324987         731    Male          <ISCED level 3A>    
12789          117    Male          <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
293942                                   Yes   
302093                                    No   
195795                                    No   
975147                                    No   
479944                                    No   
290007                                    No   
21021                                    Yes   
948592                                    No   
324987                                    No   
12789                                    Yes   

       Mother Qualifications - ISCED level 5A  \
293942                                    Yes   
302093                                     No   
195795                                     No   
975147                                     No   
479944                                     No   
290007                                    Yes   
21021                                     Yes   
948592                                     No   
324987                                    Yes   
12789                                     Yes   

       Mother Qualifications - ISCED level 5B  \
293942                                     No   
302093                                     No   
195795                                     No   
975147                                     No   
479944                                    Yes   
290007                                     No   
21021                                      No   
948592                                     No   
324987                                     No   
12789                                      No   

       Mother Qualifications - ISCED level 4  \
293942                                    No   
302093                                    No   
195795                                   Yes   
975147                                    No   
479944                                    No   
290007                                    No   
21021                                     No   
948592                                    No   
324987                                    No   
12789                                    Yes   

        Out-of-School Study Time - Homework  \
293942                                  4.0   
302093                                  4.0   
195795                                  6.0   
975147                                  3.0   
479944                                 10.0   
290007                                  6.0   
21021                                   5.0   
948592                                  5.0   
324987                                  6.0   
12789                                   3.0   

        Out-of-School Study Time - Guided Homework  \
293942                                         1.0   
302093                                         2.0   
195795                                         1.0   
975147                                         0.0   
479944                                         0.0   
290007                                         0.0   
21021                                          1.0   
948592                                         1.0   
324987                                         2.0   
12789                                          2.0   

        Out-of-School Study Time - Personal Tutor  ...  \
293942                                        0.0  ...   
302093       

In [28]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [29]:
father_edu.sample(70)

Student ID  Gender             Mother - Highest Schooling  \
908592         3630  Female                      <ISCED level 3A>    
877573        22296    Male                       <ISCED level 2>    
573831         1935    Male                  <ISCED level 3B, 3C>    
965611         1504  Female                  <ISCED level 3B, 3C>    
966319         3502  Female                  <ISCED level 3B, 3C>    
80083           558    Male                       <ISCED level 2>    
112610         3129  Female                      <ISCED level 3A>    
486409         9437  Female                       <ISCED level 2>    
110474          751    Male                      <ISCED level 3A>    
86121          4881    Male                       <ISCED level 1>    
776916         4424  Female                      <ISCED level 3A>    
831821         2780  Female                      <ISCED level 3A>    
727555         6325    Male                       <ISCED level 1>    
774451         5325  Female                       <ISCED level 1>    
865724         2622    Male                      <ISCED level 3A>    
11071          7651    Male                      <ISCED level 3A>    
669695         3227  Female                       <ISCED level 1>    
772644          939  Female                       <ISCED level 2>    
310957          213  Female                      <ISCED level 3A>    
944513         7807  Female                       <ISCED level 2>    
468490         3184    Male                      <ISCED level 3A>    
822253         8525    Male                      <ISCED level 3A>    
1008745        2739    Male                  <ISCED level 3B, 3C>    
927050         2617    Male                      <ISCED level 3A>    
22807         18180  Female                      <ISCED level 3A>    
165262         8219  Female                      <ISCED level 3A>    
835124         3266  Female                      <ISCED level 3A>    
850051         3995  Female                  <ISCED level 3B, 3C>    
32391          3464  Female                      <ISCED level 3A>    
121909          472  Female                      <ISCED level 3A>    
...             ...     ...                                    ...   
635507          178    Male                       <ISCED level 1>    
283797        10333    Male                      <ISCED level 3A>    
70085          9941    Male                      <ISCED level 3A>    
9893           3600  Female                      <ISCED level 3A>    
513552        11427  Female                      <ISCED level 3A>    
47305          4177    Male                      <ISCED level 3A>    
58400          1155    Male                  <ISCED level 3B, 3C>    
79122          3615  Female                  <ISCED level 3B, 3C>    
573381         4804  Female                      <ISCED level 3A>    
903896          250  Female                      <ISCED level 3A>    
167220         5162    Male                      <ISCED level 3A>    
811647         4715  Female                      <ISCED level 3A>    
775372           28  Female                      <ISCED level 3A>    
39763          7391  Female                  <ISCED level 3B, 3C>    
58818          2099    Male                       <ISCED level 2>    
799392         3235    Male                  <ISCED level 3B, 3C>    
13608          4801  Female                      <ISCED level 3A>    
58899          2304  Female                      <ISCED level 3A>    
33211            31    Male                      <ISCED level 3A>    
1144677        2419  Female  She did not complete <ISCED level 1>    
368282          684    Male                  <ISCED level 3B, 3C>    
812391         2688    Male                  <ISCED level 3B, 3C>    
863659         1914    Male                      <ISCED level 3A>    
499754        11133  Female                      <ISCED level 3A>    
565749         1870    Male                      <ISCED level 3A>    
76377         26498

In [30]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [31]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    84496
<ISCED level 5>    49085
<ISCED level 4>    26411
<ISCED level 2>    21955
<ISCED level 6>     6715
<ISCED level 1>     5947
<ISCED level 0>     2290
Name: Education - Father, dtype: int64

In [32]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [33]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
811744         326  Female          <ISCED level 3A>    
246324        1460  Female          <ISCED level 3A>    
993011       24112    Male          <ISCED level 3A>    
848649         378    Male      <ISCED level 3B, 3C>    
449860        3698    Male          <ISCED level 3A>    
988834       13237  Female      <ISCED level 3B, 3C>    
881401        1810    Male      <ISCED level 3B, 3C>    
281455        3915  Female          <ISCED level 3A>    
866006        3394  Female          <ISCED level 3A>    
511651        6250  Female      <ISCED level 3B, 3C>    

       Mother Qualifications - ISCED level 6  \
811744                                    No   
246324                                    No   
993011                                    No   
848649                                    No   
449860                                    No   
988834                                   Yes   
881401                                    No   
281455                                    No   
866006                                    No   
511651                                    No   

       Mother Qualifications - ISCED level 5A  \
811744                                     No   
246324                                    Yes   
993011                                     No   
848649                                     No   
449860                                     No   
988834                                    Yes   
881401                                     No   
281455                                    Yes   
866006                                     No   
511651                                     No   

       Mother Qualifications - ISCED level 5B  \
811744                                     No   
246324                                    Yes   
993011                                    Yes   
848649                                     No   
449860                                    Yes   
988834                                     No   
881401                                    Yes   
281455                                     No   
866006                                     No   
511651                                    Yes   

       Mother Qualifications - ISCED level 4  \
811744                                   Yes   
246324                                    No   
993011                                    No   
848649                                   Yes   
449860                                    No   
988834                                    No   
881401                                    No   
281455                                    No   
866006                                   Yes   
511651                                   Yes   

        Out-of-School Study Time - Homework  \
811744                                  8.0   
246324                                 10.0   
993011                                 15.0   
848649                                  1.0   
449860                                 10.0   
988834                                  3.0   
881401                                 14.0   
281455                                  6.0   
866006                                  4.0   
511651                                  0.0   

        Out-of-School Study Time - Guided Homework  \
811744                                         0.0   
246324                                         2.0   
993011                                        12.0   
848649                                         1.0   
449860                                         6.0   
988834                                         1.0   
881401                                        10.0   
281455                                         1.0   
866006                                         0.0   
511651                                         0.0   

        Out-of-School Study Time - Personal Tutor  \
811744                                        1.0   
246324                 

In [34]:
father_edu.duplicated().sum()

6947

In [35]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
father_edu_clean = father_edu.drop_duplicates(subset='Student ID', keep='last')

In [36]:
father_edu_clean.duplicated().sum()

0

In [37]:
father_edu.shape

(196899, 20)

### Mother Edu

In [38]:
mother_edu = father_edu_clean.copy()

In [39]:
mother_edu.head()

Student ID  Gender             Mother - Highest Schooling  \
1062054        6301  Female  She did not complete <ISCED level 1>    
1054542       16914    Male                      <ISCED level 3A>    
1054388       16487    Male                      <ISCED level 3A>    
1054076       15611  Female                      <ISCED level 3A>    
1053584       14175    Male                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
1062054                                    No   
1054542                                    No   
1054388                                    No   
1054076                                    No   
1053584                                    No   

        Mother Qualifications - ISCED level 5A  \
1062054                                     No   
1054542                                    Yes   
1054388                                     No   
1054076                                     No   
1053584                                    Yes   

        Mother Qualifications - ISCED level 5B  \
1062054                                     No   
1054542                                     No   
1054388                                     No   
1054076                                     No   
1053584                                     No   

        Mother Qualifications - ISCED level 4  \
1062054                                    No   
1054542                                    No   
1054388                                    No   
1054076                                    No   
1053584                                    No   

         Out-of-School Study Time - Homework  \
1062054                                  0.0   
1054542                                  0.0   
1054388                                  1.0   
1054076                                  2.0   
1053584                                  1.0   

         Out-of-School Study Time - Guided Homework  \
1062054                                         0.0   
1054542                                         0.0   
1054388                                         1.0   
1054076                                         0.0   
1053584                                         1.0   

         Out-of-School Study Time - Personal Tutor  \
1062054                                        0.0   
1054542                                        0.0   
1054388                                        0.0   
1054076                                        0.0   
1053584                                        0.0   

         Out-of-School Study Time - Commercial Company  \
1062054                                            0.0   
1054542                                            0.0   
1054388                                            0.0   
1054076                                            0.0   
1053584                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
1062054                                     1.0                        540.0   
1054542                                     0.0                        280.0   
1054388                                     1.0                        400.0   
1054076                                     0.0                        375.0   
1053584                                     1.0                        400.0   

         Learning Time - Test Language  Learning Time - Science  \
1062054                          720.0                     90.0   
1054542                          280.0                    280.0   
1054388                          400.0                    400.0   
1054076                          375.0                    375.0   
1053584                            0.0                      0.0   

         Average Math Score  Average Reading Score  Average Science Score  \
1062054           301.69028              311.29408              287.52330   
1054542           497.43728              474.25206              

In [40]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         15483
<ISCED level 3B, 3C>                      3806
<ISCED level 2>                           3612
<ISCED level 1>                            533
She did not complete <ISCED level 1>       151
Name: Mother - Highest Schooling, dtype: int64

In [41]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         63660
<ISCED level 3B, 3C>                     23596
<ISCED level 2>                          19860
<ISCED level 1>                           5189
She did not complete <ISCED level 1>      2383
Name: Mother - Highest Schooling, dtype: int64

In [42]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [43]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [44]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [45]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [46]:
mother_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
588182       10479    Male  She did not complete <ISCED level 1>    
498721        7992  Female                      <ISCED level 3A>    
481542       18148  Female                       <ISCED level 2>    
480558       15415  Female                      <ISCED level 3A>    
482114       19792  Female                      <ISCED level 3A>    
501556       16605  Female                      <ISCED level 3A>    
249706       11033    Male                      <ISCED level 3A>    
215618        1623    Male                      <ISCED level 3A>    
384013        7930  Female                      <ISCED level 3A>    
875813       17648    Male                      <ISCED level 3A>    

       <ISCED level 6>  <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
588182              No               No               No               No   
498721              No               No  <ISCED level 5>  <ISCED level 4>   
481542              No               No               No               No   
480558              No  <ISCED level 5>  <ISCED level 5>               No   
482114              No               No               No               No   
501556              No               No               No               No   
249706              No  <ISCED level 5>  <ISCED level 5>               No   
215618              No  <ISCED level 5>  <ISCED level 5>               No   
384013              No               No  <ISCED level 5>               No   
875813              No               No               No               No   

        Out-of-School Study Time - Homework  \
588182                                  2.0   
498721                                  8.0   
481542                                  2.0   
480558                                  5.0   
482114                                  2.0   
501556                                  5.0   
249706                                  3.0   
215618                                 10.0   
384013                                  4.0   
875813                                  6.0   

        Out-of-School Study Time - Guided Homework  \
588182                                         0.0   
498721                                         0.0   
481542                                         0.0   
480558                                         2.0   
482114                                         0.0   
501556                                         2.0   
249706                                         1.0   
215618                                         2.0   
384013                                         4.0   
875813                                         0.0   

        Out-of-School Study Time - Personal Tutor  ...  Average Math Score  \
588182                                        0.0  ...           376.85780   
498721                                        0.0  ...           605.39800   
481542                                        0.0  ...           558.97334   
480558                                        0.0  ...           524.23276   
482114                                        0.0  ...           552.89764   
501556                                        0.0  ...           447.81896   
249706                                        0.0  ...           642.55336   
215618                                        0.0  ...           478.35334   
384013                                        6.0  ...           516.91072   
875813                                        0.0  ...           581.32888   

        Average Reading Score  Average Science Score  Average Total Score  \
588182              329.09998              330.04476           470.953447   
498721              587.39628              613.14808           803.780120   
481542              554.67070              530.99588           734.537753   
480558              551.01686              543.77100           714.417793   
482114              533.46260              

In [47]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [48]:
mother_edu.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
13260         6758  Female                                  5.0   
70252         3384  Female                                  5.0   
38992         1496    Male                                  9.0   
210716       10067  Female                                  3.0   
61315        10889  Female                                  5.0   
82257        12012    Male                                  1.0   
32841        12492  Female                                  2.0   
123807        6333    Male                                  5.0   
141686       14170  Female                                 15.0   
143483       14207  Female                                  3.0   

        Out-of-School Study Time - Guided Homework  \
13260                                          0.0   
70252                                          1.0   
38992                                          1.0   
210716                                         2.0   
61315                                          4.0   
82257                                          0.0   
32841                                          0.0   
123807                                         0.0   
141686                                         5.0   
143483                                         3.0   

        Out-of-School Study Time - Personal Tutor  \
13260                                         8.0   
70252                                         1.0   
38992                                         0.0   
210716                                        0.0   
61315                                         2.0   
82257                                         0.0   
32841                                         0.0   
123807                                        0.0   
141686                                        1.0   
143483                                        0.0   

        Out-of-School Study Time - Commercial Company  \
13260                                             0.0   
70252                                             1.0   
38992                                            10.0   
210716                                            0.0   
61315                                             3.0   
82257                                             0.0   
32841                                             0.0   
123807                                            0.0   
141686                                            0.0   
143483                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
13260                                      0.0                        240.0   
70252                                      3.0                        240.0   
38992                                      1.0                        180.0   
210716                                     0.0                        300.0   
61315                                      2.0                        300.0   
82257                                      0.0                        320.0   
32841                                      0.0                        300.0   
123807                                     0.0                        250.0   
141686                                     5.0                        400.0   
143483                                     3.0                        240.0   

        Learning Time - Test Language  Learning Time - Science  \
13260                           240.0                    240.0   
70252                           240.0                    300.0   
38992                           225.0                    360.0   
210716                          300.0                    300.0   
61315                           600.0                    120.0   
82257                           320.0                    320.0   
32841                           300.0                    300.0   
123807                          180.0                     90.0   

In [49]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [50]:
mother_edu.sample(70)

Student ID  Gender  Out-of-School Study Time - Homework  \
180631         892  Female                                 28.0   
22167         3985  Female                                  6.0   
17553         2697    Male                                  6.0   
12203         9656    Male                                  3.0   
175049        6500    Male                                  1.0   
119165       22003  Female                                  2.0   
144076       10290    Male                                  1.0   
18619         5888  Female                                  3.0   
176185        5879  Female                                  7.0   
186204        4768  Female                                  8.0   
23456         3960    Male                                  0.0   
23481         4275    Male                                  3.0   
168406       18442  Female                                 10.0   
17282          894    Male                                  9.0   
4898         14047  Female                                  5.0   
167103       16008    Male                                  2.0   
171606       24893    Male                                  3.0   
185704        3380  Female                                  1.0   
88822        12238  Female                                  2.0   
21816         4139  Female                                 20.0   
118456       15756    Male                                  4.0   
91734        10333    Male                                  2.0   
59354         9169    Male                                  2.0   
22762         2512  Female                                  1.0   
15759         1068  Female                                  8.0   
185778        1783  Female                                  5.0   
4440         21791  Female                                  8.0   
66651          992  Female                                 12.0   
10391        23969  Female                                  2.0   
169363       27167  Female                                  2.0   
...            ...     ...                                  ...   
18906         9094  Female                                  3.0   
16120         4058  Female                                  1.0   
10574         6066    Male                                  2.0   
198521        9214  Female                                  6.0   
186205        4654  Female                                  6.0   
2198         18786    Male                                 25.0   
61073        18613  Female                                 28.0   
204312        1564  Female                                  0.0   
86717         6464  Female                                  4.0   
5937         10277  Female                                  4.0   
204648        6296    Male                                  5.0   
144664       19846  Female                                 10.0   
9418         21013  Female                                  4.0   
59207        13812    Male                                  6.0   
182139       17535  Female                                  8.0   
22919        27285  Female                                  8.0   
12579         7984  Female                                  2.0   
204225        1913  Female                                  6.0   
80247        16911  Female                                 10.0   
174699       14657    Male                                  2.0   
13923        20130    Male                                  6.0   
76698         7731    Male                                  1.0   
13593        12287    Male                                  5.0   
15948         4934    Male                                  1.0   
81483         7508  Female                                  2.0   
179234       11049  Female                                 22.0   
176634       10752    Male                                  9.0   
183580        5951  Female                             

In [51]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [52]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    19289
<ISCED level 5>    14586
<ISCED level 4>     5106
<ISCED level 2>     3612
<ISCED level 6>     2630
<ISCED level 1>      533
<ISCED level 0>      151
Name: Education - Mother, dtype: int64

In [53]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [54]:
mother_edu.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
92192         3404  Female                                  4.0   
175521       23021  Female                                 30.0   
177269       10188    Male                                 10.0   
183116       10497    Male                                  1.0   
111257       14464    Male                                  8.0   
63453        11340    Male                                  5.0   
106481        9987    Male                                  4.0   
186504        8791  Female                                 14.0   
180679        3436  Female                                  4.0   
204807        3958    Male                                  4.0   

        Out-of-School Study Time - Guided Homework  \
92192                                          0.0   
175521                                         4.0   
177269                                         0.0   
183116                                         1.0   
111257                                         3.0   
63453                                          2.0   
106481                                         0.0   
186504                                         0.0   
180679                                         1.0   
204807                                         0.0   

        Out-of-School Study Time - Personal Tutor  \
92192                                         0.0   
175521                                        4.0   
177269                                        0.0   
183116                                        0.0   
111257                                        3.0   
63453                                         2.0   
106481                                        0.0   
186504                                        2.0   
180679                                        1.0   
204807                                        0.0   

        Out-of-School Study Time - Commercial Company  \
92192                                             8.0   
175521                                            0.0   
177269                                            0.0   
183116                                            0.0   
111257                                            2.0   
63453                                             2.0   
106481                                            0.0   
186504                                            0.0   
180679                                            0.0   
204807                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
92192                                      0.0                        240.0   
175521                                     1.0                        250.0   
177269                                     0.0                        375.0   
183116                                     1.0                        120.0   
111257                                     0.0                        180.0   
63453                                      0.0                        165.0   
106481                                     0.0                        300.0   
186504                                     0.0                        240.0   
180679                                     2.0                        225.0   
204807                                     0.0                        180.0   

        Learning Time - Test Language  Learning Time - Science  \
92192                           240.0                      0.0   
175521                          200.0                    150.0   
177269                          375.0                    375.0   
183116                          180.0                    120.0   
111257                          150.0                    360.0   
63453                           220.0                    165.0   
106481                          300.0                    300.0   
186504                          240.0                      0.0   

In [55]:
mother_edu['Student ID'].duplicated().sum()

22322

In [56]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [57]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
mother_edu_clean = mother_edu.drop_duplicates(subset='Student ID', keep='last')

In [58]:
mother_edu['Student ID'].duplicated().sum()

22322

In [59]:
mother_edu.shape

(45907, 16)

In [60]:
mother_edu_clean.shape

(23585, 16)

In [61]:
mother_edu_clean.duplicated().sum()

0

In [62]:
mother_edu_clean.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
175153       24759    Male                                  8.0   
105266       10096    Male                                  2.0   
188507        3618  Female                                  9.0   
156715        1813    Male                                  3.0   
170463       14420  Female                                 10.0   
64541         4521    Male                                  4.0   
143744       20491    Male                                 18.0   
165740        9991  Female                                  4.0   
101770       13846    Male                                  1.0   
143327       27908  Female                                  8.0   

        Out-of-School Study Time - Guided Homework  \
175153                                         0.0   
105266                                         1.0   
188507                                         0.0   
156715                                         0.0   
170463                                         4.0   
64541                                          0.0   
143744                                         4.0   
165740                                         0.0   
101770                                         0.0   
143327                                         0.0   

        Out-of-School Study Time - Personal Tutor  \
175153                                        0.0   
105266                                        0.0   
188507                                        0.0   
156715                                        0.0   
170463                                        2.0   
64541                                         0.0   
143744                                        0.0   
165740                                        0.0   
101770                                        0.0   
143327                                        6.0   

        Out-of-School Study Time - Commercial Company  \
175153                                            0.0   
105266                                            0.0   
188507                                            0.0   
156715                                            0.0   
170463                                            0.0   
64541                                             0.0   
143744                                            0.0   
165740                                            0.0   
101770                                            0.0   
143327                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
175153                                     0.0                        240.0   
105266                                     1.0                        225.0   
188507                                     3.0                        240.0   
156715                                     0.0                        150.0   
170463                                     4.0                        375.0   
64541                                      0.0                        200.0   
143744                                     0.0                        300.0   
165740                                     0.0                        300.0   
101770                                     0.0                        375.0   
143327                                     0.0                        240.0   

        Learning Time - Test Language  Learning Time - Science  \
175153                          240.0                    180.0   
105266                          270.0                    180.0   
188507                          240.0                    320.0   
156715                          150.0                    100.0   
170463                          375.0                    375.0   
64541                           175.0                    140.0   
143744                          240.0                    240.0   
165740                          300.0                    240.0   

In [63]:
pisa_clean = mother_edu_clean.copy()

In [64]:
pisa_clean.shape

(23585, 16)

In [65]:
pisa_clean = mother_edu.copy()

In [66]:
pisa_clean.shape

(45907, 16)

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [67]:

# pisa_clean.replace(["NaN"], np.nan, inplace = True)
# pisa_clean.dropna(inplace = True)

In [68]:
# Check to see if null values correctly dropped
pisa_clean

Student ID  Gender  Out-of-School Study Time - Homework  \
235845        1647    Male                                  2.0   
212618       18805  Female                                  0.0   
212616       18913    Male                                  2.0   
212606       12822    Male                                 30.0   
212595        9143    Male                                  4.0   
212592       10479    Male                                  2.0   
212583       16295  Female                                  1.0   
212580       16142  Female                                  5.0   
212577       17761    Male                                  1.0   
212576       17846    Male                                  2.0   
212571       17414  Female                                  2.0   
212563       14327  Female                                  1.0   
212554       13635    Male                                  2.0   
212546       15437  Female                                  8.0   
212534       15337    Male                                  2.0   
212510       21333    Male                                  7.0   
212635       14280    Male                                  3.0   
212401       29985    Male                                  0.0   
212668       19354  Female                                  5.0   
212771       22336    Male                                  1.0   
213577       22804  Female                                  5.0   
213433       24651  Female                                  7.0   
213049       12825    Male                                  2.0   
213047       13375    Male                                  1.0   
213044       13203  Female                                 10.0   
213029       13900    Male                                  3.0   
212952       11498    Male                                  7.0   
212950       10904  Female                                  9.0   
212876       10648  Female                                  5.0   
212867       25210  Female                                 10.0   
...            ...     ...                                  ...   
42908        14502  Female                                 11.0   
42911        14461    Male                                 15.0   
42912        14415    Male                                 10.0   
42913        14413  Female                                 12.0   
42914        14509  Female                                  3.0   
42916        14793    Male                                  3.0   
42917        15089  Female                                 10.0   
42919        15082    Male                                 10.0   
42892        15992  Female                                  2.0   
42890        16106    Male                                  4.0   
42889        16146  Female                                  5.0   
42885        16256    Male                                  5.0   
42852        12828    Male                                  6.0   
42853        12566    Male                                 28.0   
42855        12785  Female                                  1.0   
42856        12768  Female                                  3.0   
42859        12665    Male                                 16.0   
42860        12624  Female                                  0.0   
42861        12804    Male                                  3.0   
42845        14395  Female                                  1.0   
42863        14401    Male                                  5.0   
42868        15545  Female                                  8.0   
42869        15544    Male                                  8.0   
42871        15718  Female                                 10.0   
42872        15537    Male                                  0.0   
42873        15456  Female                                  7.0   
42880        15785  Female                                 14.0   
42884        16271    Male                             

In [69]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(45907, 16)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [70]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [71]:
list(pisa_clean)

['Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [72]:
pisa_clean

Student ID  Gender  Out-of-School Study Time - Homework  \
235845        1647    Male                                  2.0   
212618       18805  Female                                  0.0   
212616       18913    Male                                  2.0   
212606       12822    Male                                 30.0   
212595        9143    Male                                  4.0   
212592       10479    Male                                  2.0   
212583       16295  Female                                  1.0   
212580       16142  Female                                  5.0   
212577       17761    Male                                  1.0   
212576       17846    Male                                  2.0   
212571       17414  Female                                  2.0   
212563       14327  Female                                  1.0   
212554       13635    Male                                  2.0   
212546       15437  Female                                  8.0   
212534       15337    Male                                  2.0   
212510       21333    Male                                  7.0   
212635       14280    Male                                  3.0   
212401       29985    Male                                  0.0   
212668       19354  Female                                  5.0   
212771       22336    Male                                  1.0   
213577       22804  Female                                  5.0   
213433       24651  Female                                  7.0   
213049       12825    Male                                  2.0   
213047       13375    Male                                  1.0   
213044       13203  Female                                 10.0   
213029       13900    Male                                  3.0   
212952       11498    Male                                  7.0   
212950       10904  Female                                  9.0   
212876       10648  Female                                  5.0   
212867       25210  Female                                 10.0   
...            ...     ...                                  ...   
42908        14502  Female                                 11.0   
42911        14461    Male                                 15.0   
42912        14415    Male                                 10.0   
42913        14413  Female                                 12.0   
42914        14509  Female                                  3.0   
42916        14793    Male                                  3.0   
42917        15089  Female                                 10.0   
42919        15082    Male                                 10.0   
42892        15992  Female                                  2.0   
42890        16106    Male                                  4.0   
42889        16146  Female                                  5.0   
42885        16256    Male                                  5.0   
42852        12828    Male                                  6.0   
42853        12566    Male                                 28.0   
42855        12785  Female                                  1.0   
42856        12768  Female                                  3.0   
42859        12665    Male                                 16.0   
42860        12624  Female                                  0.0   
42861        12804    Male                                  3.0   
42845        14395  Female                                  1.0   
42863        14401    Male                                  5.0   
42868        15545  Female                                  8.0   
42869        15544    Male                                  8.0   
42871        15718  Female                                 10.0   
42872        15537    Male                                  0.0   
42873        15456  Female                                  7.0   
42880        15785  Female                                 14.0   
42884        16271    Male                             

In [73]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45907 entries, 235845 to 23640
Data columns (total 18 columns):
Student ID                                       45907 non-null int64
Gender                                           45907 non-null object
Out-of-School Study Time - Homework              45907 non-null float64
Out-of-School Study Time - Guided Homework       45907 non-null float64
Out-of-School Study Time - Personal Tutor        45907 non-null float64
Out-of-School Study Time - Commercial Company    45907 non-null float64
Out-of-School Study Time - With Parent           45907 non-null float64
Learning Time - Mathematics                      45907 non-null float64
Learning Time - Test Language                    45907 non-null float64
Learning Time - Science                          45907 non-null float64
Average Math Score                               45907 non-null float64
Average Reading Score                            45907 non-null float64
Average Science Score    

In [74]:
pisa_clean.shape

(45907, 18)

<a id='sav'></a>
## Store

In [75]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References